### ISSUE TITLE
BlenderBIM: Changes to blender material are reflected ifc object as well

### ISSUE THREAD

**Issue by theoryshaw** [IfcOpenShell/IfcOpenShell#1585]
If an IfcMaterial is applied to a wall through the IfcMateralLayerSet, it would nice if you changed the blender material attribute such as color, that the wall would change color too. 

video example: https://www.dropbox.com/s/p8sx68xewojgv06/2021-07-24_15-38-39.mp4?dl=0

**Comment by Moult** [comment]
Yes, the question is which layer would apply? The first? The last?

**Comment by theoryshaw** [comment]
Defer to you. But would guess, the 1st IfcMaterialLayer, as hopefully, in the future, we'll be able to visualize and translate multiple layers into Blender geometry.

**Comment by Moult** [comment]
I had a start at looking at this, but it's a little bit trickier. In IFC, it is possible to have a material assigned with _no_ colour. If I make this change, then users will expect, upon authoring, that whatever colour they see in their scene is present in the IFC. So right now, whenever you sync your 3D geometry, it also syncs the colours that you have applied (i.e. the materials in that materials tab in Blender which you're looking at).

The current behaviour, where you can add a material via the IFC Object Material tab and never create a colour is by design to allow this distinction, but obviously as you are experiencing it is also a bit unintuitive.

I'll see what I can do. For the moment, at least I now support importing material fallbacks which is a start (i.e. objects with material styles, but no other shape-based styles).

**Comment by theoryshaw** [comment]
sounds good. 

Similar to [this](https://github.com/IfcOpenShell/IfcOpenShell/issues/1592) issue, could there be a preference set somewhere where the ifcmaterial and ifcsurface styles are interconnected (bi-directional) with blender material attributes?  

That is, change the color in the ifcsurface style, and it automatically changes the base color in the blender materials, or vice versa, change the base color in the blender material and it changes the ifcsurface style color.

Just thinking out loud.

**Comment by Moult** [comment]
Fixed and fixed. Note that there is no such thing as a bi-directional link for colours - colours are only defined in one location: a surface style, and nowhere else.

The first fix auto-adds Blender materials when you add object materials even if it's technically not necessary, simply because you're right, it is just a lot more intuitive if it does so.

The second fix auto syncs colour changes. Use the "diffuse colour" down in viewport display to see it in action. Note that this is the first time I've built the ability for materials to be auto-synced (in the past, I was slowly building auto syncs for object editing, object locations, and object names) ... but now this is the first material-based syncing system, so keep a keen eye out for bugs.
